openedx/openedx-platform
# [BD-04] Convert SplitTest XModule to XBlock. [SE-3640]

Converts the SplitTest XModule into an XBlock.

Part of [XModule to XBlock Conversion work](https://openedx.atlassian.net/wiki/spaces/AC/pages/1472790755/XModule+to+XBlock+Conversion). 

**Testing instructions**:

* In a new course set up a SplitTest component. In the UI it is called [Content Experiments](https://edx.readthedocs.io/projects/edx-partner-course-staff/en/latest/course_features/content_experiments/content_experiments_overview.html).
* Interact with it in LMS to check its behavior.
* Export the course.
* Checkout this branch.
* Run:
 ```
 make lms-shell
 pip install -e common/lib/xmodule
 exit
 // In case of webpack errors run make lms-static
 make lms-restart
 make studio-restart
 ```
* Check that the component behaves as expected in Studio and LMS.
* Import the course archive in a new course.
* Check that the component in the new course behaves as expected in Studio and LMS.

**MRO Analysis**:

Descriptor MRO | Module MRO | Block MRO | Notes
------------ | ------------- | ---------------- | ------------
<class 'xmodule.split_test_module.SplitTestDescriptor'> | <class 'xmodule.split_test_module.SplitTestModule'> | <class 'xmodule.split_test_module.SplitTestBlock'> |
<class 'xmodule.split_test_module.SplitTestFields'> | <class 'xmodule.split_test_module.SplitTestFields'> | <class 'xmodule.split_test_module.SplitTestFields'> |
<class 'xmodule.seq_module.SequenceDescriptor'> | | | All needed but not already overridden attributes have been set on the SplitTestBlock (has_author_view, non_editable_metadata_fields).
<class 'xmodule.seq_module.SequenceMixin'> | | <class 'xmodule.seq_module.SequenceMixin'> |
<class 'xmodule.seq_module.SequenceFields'> | | <class 'xmodule.seq_module.SequenceFields'> |
<class 'xmodule.seq_module.ProctoringFields'> | | <class 'xmodule.seq_module.ProctoringFields'> |
<class 'xmodule.mako_module.MakoModuleDescriptor'> | | | Inherits from (MakoTemplateBlockBase, XModuleDescriptor) and has the get_html() method which is not needed anymore.
<class 'xmodule.mako_module.MakoTemplateBlockBase'> | | <class 'xmodule.mako_module.MakoTemplateBlockBase'> |
<class 'xmodule.xml_module.XmlDescriptor'> | | | Doesn't have any attributes.
<class 'xmodule.xml_module.XmlMixin'> | | <class 'xmodule.xml_module.XmlMixin'> |
<class 'xmodule.xml_module.XmlParserMixin'> | | <class 'xmodule.xml_module.XmlParserMixin'> |
<class 'xmodule.x_module.XModuleDescriptor'> | | |
<class 'xmodule.x_module.XModuleDescriptorToXBlockMixin'> | | <class 'xmodule.x_module.XModuleDescriptorToXBlockMixin'> |
 | <class 'xmodule.x_module.XModule'> | |
 | <class 'xmodule.x_module.XModuleToXBlockMixin'> | <class 'xmodule.x_module.XModuleToXBlockMixin'> |
<class 'xmodule.x_module.HTMLSnippet'> | <class 'xmodule.x_module.HTMLSnippet'> | | This is only needed for compiling JS+CSS for the block but we are using the Sequence's js so don't need that.
<class 'xmodule.x_module.ResourceTemplates'> | | <class 'xmodule.x_module.ResourceTemplates'> |
<class 'xmodule.x_module.XModuleMixin'> | <class 'xmodule.x_module.XModuleMixin'> | <class 'xmodule.x_module.XModuleMixin'> |
<class 'xmodule.x_module.XModuleFields'> | <class 'xmodule.x_module.XModuleFields'> | <class 'xmodule.x_module.XModuleFields'> |
<class 'xblock.core.XBlock'> | <class 'xblock.core.XBlock'> | <class 'xblock.core.XBlock'> |
